File: elixir/accumulate/accumulate.exs
defmodule Accumulate do
  @doc """
    Given a list and a function, apply the function to each list item and
    replace it with the function's return value.

    Returns a list.

    ## Examples

      iex> Accumulate.accumulate([], fn(x) -> x * 2 end)
      []

      iex> Accumulate.accumulate([1, 2, 3], fn(x) -> x * 2 end)
      [2, 4, 6]

  """

  @spec accumulate(list, (any -> any)) :: list
  def accumulate(list, fun) do
    do_accumlate(list, fun, [])
  end

  defp do_accumlate([], _, acc), do: Enum.reverse(acc)
  defp do_accumlate([head | tail], fun, acc) do
    do_accumlate(tail, fun, [fun.(head) | acc])
  end
end

File: java/accumulate/src/main/java/Accumulate.java
import java.util.LinkedList;
import java.util.List;
import java.util.function.Function;

public class Accumulate {

  public static <T> List<T> accumulate(List<T> items, Function<T, T> mapper) {
    List<T> mapped = new LinkedList<>();
    for (T item : items) {
      mapped.add(mapper.apply(item));
    }
    return mapped;
  }

}


